python/setuputils.py

```diff
@@ -46,7 +68,121 @@ def clean_folder(path):
             os.remove(file)
 
 
-def clone_repo(name, GIT_REPOSITORY, GIT_TAG, force_clone=False):
+def use_raft_package(raft_path, cpp_build_path,
+                     git_info_file='../cpp/cmake/Dependencies.cmake'):
+    """
+    Function to use the python code in RAFT in package.raft
+
+    - If RAFT symling already exists, don't change anything. Use setup.py clean
+        if you want to change RAFT location.
+    - Uses RAFT located in $RAFT_PATH if $RAFT_PATH exists.
+    - Otherwise it will look for RAFT in the libcuml build folder,
+        located either in the default location ../cpp/build or in
+        $CUML_BUILD_PATH.
+    -Otherwise it will clone RAFT into _external_repositories.
+        - Branch/git tag cloned is located in git_info_file in this case.
+
+    """
+    if not os.path.islink('cuml/raft'):
+        if not raft_path:
+            raft_path, raft_cloned = \
+                clone_repo_if_needed('raft', cpp_build_path,
+                                     git_info_file=git_info_file)
+
+            raft_path = '../' + raft_path
+
+        else:
+            print("-- Using RAFT_PATH variable, RAFT found at " +
+                  str(os.environ['RAFT_PATH']))
+            raft_path = os.environ['RAFT_PATH']
+
+        try:
+            os.symlink(raft_path + 'python/raft', 'cuml/raft')
+        except FileExistsError:
+            os.remove('cuml/raft')
+            os.symlink(raft_path + 'python/raft', 'cuml/raft')
+
+    else:
+        print("-- Using already existing RAFT folder, source located at")
+        print(os.path.realpath('cuml/raft'))
+
+
+def clone_repo_if_needed(name, cpp_build_path,
+                         git_info_file='../cpp/cmake/Dependencies.cmake'):
+    if cpp_build_path:
+        cpp_build_path = '../' + cpp_build_path
+    else:
+        cpp_build_path = '../cpp/build/'
+
+    repo_cloned = get_submodule_dependency(name,
+                                           cpp_build_path=cpp_build_path,
+                                           git_info_file=git_info_file)
+
+    if repo_cloned:
+        repo_path = '_external_repositories/' + name + '/'
+    else:
+        repo_path = os.path.join(cpp_build_path, name + '/src/' + name + '/')
+
+    return repo_path, repo_cloned
+
+
+def get_submodule_dependency(repo,
```

Comment:
In general, I'm not too comfortable with parsing cmake files inside python setup.py. You did mention the reason for doing this cloning in python sometime ago, @dantegd and I seem to have forgotten it now! Can you please (re)help me remember the reason for doing this, please?

Reply:
It is for cloning RAFT if needed by Python (or any other repo), and having only one central place where RAFT is pinned. Otherwise we would need to have 2 places where the pinned commit of RAFT is defined, introducing potential conflicts. 

Typically, RAFT will only be cloned by cmake and python uses that RAFT clone, or RAFT will be located in the location pointed by RAFT_PATH and both cmake and python use it from there. But if neither is true and a developer is only building the Python package, then python clones it itself (this was the same situation with treelite and other things that are/were needed by the python package)

Reply:
To solve the need to parse cmake from setup.py and decouple things, we could create a config file that both cmake and setup.py read and is clearer/easier for devs to modify, what do you think?

Reply:
Ideally, none of the treelite/faiss/spdlog dependencies should have been exposed to the users who'd only want to build python code (but have libcuml++.so and cpp/include headers installed already)! So, clearly, this seems to be an issue with our C++ layer. I have filed issue #2138 to address this one.